### PR TITLE
Adding code to verify datastore engine according to URI

### PR DESCRIPTION
**Pull request by Himanshu372**
Adding code to verify datastore engine is in line with URI provided in configurations

### Added 
1. Added a `DatastoreEngine` type which takes up const permissible engine type values
2. Added helper methods for the type `DatastoreEngine`
3. Added `verifyDatastoreEngine` func to handle verification of engine in line with URI 
4. Added tests 

## References

Addresses issue - https://github.com/openfga/openfga/issues/681

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [x] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

**Comment by adriantam** [comment]
@Himanshu372 - converted as draft for you.  Feel free to mark it as "Ready for review" when you are ready :)

File: cmd/run/run.go [hunk 1]
```diff
@@ -355,7 +370,31 @@ func ReadConfig() (*Config, error) {
 	return config, nil
 }
 
+func verifyDatastoreEngine(URI string, Engine DatastoreEngine) error {
```
Comment (jon-whit) [inline cmd/run/run.go:373]:
```suggestion
func verifyDatastoreEngine(uri string, engine DatastoreEngine) error {
```

File: cmd/migrate.go [hunk 2]
```diff
@@ -64,6 +65,10 @@ func runMigration(_ *cobra.Command, _ []string) error {
 	targetVersion := viper.GetUint(versionFlag)
 	timeout := viper.GetDuration(timeoutFlag)
 
+	if err := run.VerifyDatastoreEngine(uri, run.DatastoreEngine(engine)); err != nil {
```
Comment (jon-whit) [inline cmd/migrate.go:68]: `migrate` package shouldn't have to depend on the `run` package. They are two separate entrypoints. Let's relocate the shared utilities to the cmd utilities package.

Comment (Himanshu372) [reply cmd/migrate.go]: Refactored shared code to `exec_common` folder

**Comment by Himanshu372** [comment]
@adriantam @jon-whit  Any feedback on this PR?

File: cmd/exec_common/shared.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	}
+	return &validEngine, nil
+}
+
+func VerifyDatastoreEngine(uri string, engine DatastoreEngine) error {
+	var URIType string
+	URITokens := strings.Split(uri, ":")
+	if len(URITokens) > 1 {
+		URIType = URITokens[0]
+		switch URIType {
```
Comment (jon-whit) [inline cmd/exec_common/shared.go:42]: These uris are not correctly defined. We have `postgres://...` and `mysql://...`, not `postgresql://` and `mysqlx`

Please also fix the corresponding tests in `run/run_test.go`.

Comment (jon-whit) [reply cmd/exec_common/shared.go:42]: `./openfga run --datastore-engine mysql --datastore-uri 'postgres://postgres:password@localhost:5432/postgres'` should return this validation error but currently does not as a result of this.

Also, I'm confused by the `mysqlx://` bit.. If you choose to run OpenFGA with MySQL, then your datastore connection settings wouldn't involve a datastore URI.

`./openfga run --datastore-engine mysql --datastore-uri 'root:secret@tcp(localhost:3306)/openfga'`

Are we intending to change the overall CLI behavior to require a datastore uri scheme as a prefix? It seems that is what would be required to make this functionality work as intended.

Comment (Himanshu372) [reply cmd/exec_common/shared.go]: `mysqlx` was a mistake on my end, which I have amended. It is applicable to `X` plugin, which I dont so is supported. 

Also about your 2nd comment, which was regarding `mysql` engine not requiring `mysql*` type URI is handled and I fixed tests too 

Let me know if you think any more additions are required @jon-whit

**Comment by Himanshu372** [comment]
@jon-whit Thanks for replying. Sure, I can do a different PR with enum change. I am not sure about behaviour with viper, I can take a look and try.

**Comment by jon-whit** [comment]
> @jon-whit Thanks for replying. Sure, I can do a different PR with enum change. I am not sure about behaviour with viper, I can take a look and try.

Awesome! Thank you very much. Closing this for now 👍

**Pull request closed without merging**